### DISCUSSION TITLE
Where does scholarly increment to the next page of search results (in a pubs search)?

### DISCUSSION THREAD

**Discussion by stanleyrhodes** [scholarly-python-package/scholarly#429] [Q&A] [answered]
Where does this package control going from page to page of results? I've been chasing the code up and down and I can't figure it out. When calling `search_pubs()` the `start_index` defaults to 0. But I've been through the whole code, especially `_construct_url()`, `search_publications()`, and `_SearchScholarIterator()`, and I can't find anything that looks like it either increments `start_index` to construct a new URL, or somehow updates the URL itself while looping through all the results.

The reason I want to know, besides understanding the package overall, is that I would like to script a wait-time between page requests (and later, `fill` requests) so GS doesn't become upset about the traffic. That's in addition to using ProxyManager, which I have set up and working. I've been laboring to understand the inner workings of this package for a while, but intensely in the last month, because I need to use `search_pubs()` for a quantitative literature review of thousands of papers. GS and Scholarly are the only game in town that can do it. I've been using Python for years, but I'm not very familiar with this depth of OOP with so many nested methods. I must have read the code that goes to the next page, but not have recognized it as such.

**Comment by arunkannawadi** [comment, marked as answer]
Hi Stanley, look at this implementation of `__next__` here: https://github.com/scholarly-python-package/scholarly/blob/main/scholarly/publication_parser.py#L80-L93 The iterator is not user-facing, but manages fetching results from multiple pages. You might want to add a delay in the `elif` block there. Overall, I think that'd be a good feature to have, so feel free to submit a pull request with the delay that can be specified by the user.

**Reply by stanleyrhodes** [reply]
Thanks Arun, I had seen that but I didn't realize what it was doing. Now I understand it is taking the URL associated with the Next button and icon.

I would like to contribute but I am not sure the best way to add in the delay. I'm still getting a hang of the OOP here. I do not wish to hardcode it within the `elif`, so that, as you say, the user can specify it. Would you put the parameter in `search_pubs()` to then be passed down the line to `_SearchScholarIterator()`?

Related to including a delay feature, I have another small problem. I do not want the duration of sleep to be the same every time. I've already been working on this, and I have a function that approximates human behavior browsing time for pages, based on data from my own browsing behavior. This function samples a gamma distribution + constant that should be close enough to real browsing times that it doesn't cause objectionable traffic (of course, this will be slower). Like so:

```
import numpy as np 
import time

def generate_page_time():
    result = np.random.gamma(4.5, 1.25, 1)
    result = result.__add__(2.5)
    result = np.round_(result, decimals = 3)
    result = result[0]
    return result

print('About to sleep...')
time.sleep(generate_page_time())
print('Ok, awake now!')
```

What would be the preferred way of using this? For example, I could have a `delay_function` parameter that is `None` by default as well as a a `delay` (`int` or maybe `float`) parameter that is `None` by default for `search_pubs()`. In `__next___()` I could check for the presence of either of these before executing the rest of the code in the routine and delay accordingly. This would let people create their own delay function and call it. Let me know if this is sensible, and whether/how you'd do it differently. I'd like a little guidance to make sure the contribution is solid.

**Reply by stanleyrhodes** [reply]
Ok, here is the design I mentioned, placed in modified scholarly code that doesn't have to talk to GS. In those places where it would be scraping pages, it uses a test array instead. I should be able to revise the real codebase to do this. Let me know if you see any concerns. One concern I have is that I'm not sure the best way to use the type system to require a function as a parameter, but also require that the value returned by the function is float.

```python
import numpy as np 
import time

def generate_page_time():
    result = np.random.gamma(4.5, 1.25, 1)
    result = result.__add__(2.5)
    result = np.round_(result, decimals = 3)
    result = result[0]
    return result

# Code from scholarly modified with foo suffixes.

class _SearchScholarIteratorfoo(object):
    def __init__(self, navfoo, urlfoo: str,
                 delay_in_seconds: float = None,
                 delay_from_function: any = None):
        self._urlfoo = urlfoo
        self._load_urlfoo(urlfoo)
        self._navfoo = navfoo
        self._delay_in_seconds = delay_in_seconds
        self._delay_from_function = delay_from_function
    
    def _load_urlfoo(self, urlfoo: str):
        self._somevar_foo = urlfoo
        self._posfoo = 0
        # use test array
        self._rows = ["Ford", "Volvo", "BMW", "Mercedes"]

    def __iter__(self):
        return self

    def __next__(self):
        seconds = 0
        fseconds = 0
        # If either of the delays are present, use them.
        # A user COULD use both at once, which seems fine.
        if self._delay_in_seconds is not None:
            seconds = self._delay_in_seconds
            time.sleep(seconds)
        if self._delay_from_function is not None:
            fseconds = self._delay_from_function()
            time.sleep(seconds)
        # provide some test output
        if self._posfoo < len(self._rows):
            row = (self._rows[self._posfoo] + 
                   "&" + str(seconds) + 
                   "&" + str(fseconds) + 
                   "&" + self._urlfoo)
            self._posfoo += 1
            return row
        else:
            raise StopIteration

class _Scholarlyfoo:
    """Class that manages the API for scholarly"""

    def __init__(self):
        self.__navfoo = Navigatorfoo()

    def search_pubsfoo(self, urlfoo: str,
                       delay_in_seconds: float = None,
                       delay_from_function: any = None
                       )->_SearchScholarIteratorfoo:

        return self.__navfoo.search_publicationsfoo(urlfoo,
                                                    delay_in_seconds,
                                                    delay_from_function)

class Navigatorfoo(object):
    """A class used to navigate pages on google scholar."""

    def __init__(self):
        self._max_retriesfoo = 5

    def search_publicationsfoo(self, urlfoo: str,
                               delay_in_seconds: float = None,
                               delay_from_function: any = None
                               ) ->_SearchScholarIteratorfoo:
        """Returns a Publication Generator given a url
        :param url: the url where publications can be found.
        :type url: str
        :returns: An iterator of Publications
        :rtype: {_SearchScholarIterator}
        """
        return _SearchScholarIteratorfoo(self, urlfoo,
                                         delay_in_seconds,
                                         delay_from_function)

scholarlyfoo = _Scholarlyfoo()

testfoo = scholarlyfoo.search_pubsfoo(urlfoo='a test url string thing',
                                      # delay_from_function=generate_page_time,
                                      delay_in_seconds=.5)

for item in testfoo:
    print(str(item))
```

**Reply by arunkannawadi** [reply]
You introduced a delay before the `if` block that is unnecessary. If `self._pos < len(self._rows)` holds true, it means that we are creating publication objects from a page that is already loaded, there's no delay needed there since we're not requesting anything from GS. If that fails, then the iterator looks for the next button and see if it's active. If it's not disabled, then it fetches the next page. That's where we might want a delay.

Since you've created a function that mocks up human behavior, we can have a boolean variable passed through `search_pubs` all the way down to `_SearchScholarIterator` that can either choose to have this delay or not.

**Reply by arunkannawadi** [reply]
Feel free to open a PR with your preliminary design and it's easier to suggest modifications or comment on your contribution. Here's the section on how to contribute: https://github.com/scholarly-python-package/scholarly/blob/main/.github/CONTRIBUTING.md#how-to-contribute

**Reply by stanleyrhodes** [reply]
Yes, thanks for pointing out the sleep placements. I did simplify the code a bit too much in my example for convenience sake, but I would put it before the next button check. I can see the benefit of having a boolean, but also the value of users being able to easily customize their delay timing. Working more on the project that necessitated including this delay may give me a better idea of what I think would better serve the intended users. I don't know without some experimentation whether, say, the kind of proxy being used changes the delay time preference. As you said, we can discuss with a PR and I'll have more experience by then.

Thanks also for linking me to the contribution how-to. To be honest, I don't have the time right now to fulfill all the conditions in that list. Particularly figuring out how to modify or create new test cases, as I haven't done that before outside a very basic tutorial I took a few years ago. Once the draft paper--the one I'm using scholarly for--is in internal review I'll circle back and give it all a shot using your checklist. That may be closer to the end of the year. But I'll do it before journal submission, when I'm making sure the results are replicable and citing the software version.

I appreciate you taking the time to give me some guidance! This--scholarly--is a really important project.

**Reply by arunkannawadi** [reply]
No worries. I might take the snippet you posted and make you a co-author on those commits in the mean time. Great to see this library being used scholarly work.

**Reply by stanleyrhodes** [reply]
Sure, of course I'm fine with that if you think that's worthwhile. I still hope to run through making test cases in the future. I made the delay code updates and pushed the changes to the scholarly repo fork in my Github repos. The changes are untested and I'm not sure they're working properly. I need to see if I can get the logging working and have a more verbose, play-by-play account of what's happening when I run search_pubs(). I can say, unfortunately, I had no luck with search_pubs() before making those changes, and I've had no luck afterward. GS doesn't seem to like Brightdata's data center proxies either; if anything, it's worse now than it was a couple months ago.

I think I should start a discussion in a new thread about what we (users) can realistically expect from search_pubs() right now, how people have used it to get results, dos and don'ts with proxies, and so on. It may be that I'm doing a lot of things wrong, or it may be that GS is just too good at blocking anything and everything search_pubs().
